File: Nian_Karl/QL/src/ast/visitor/Evaluator.java
package ast.visitor;

import ast.expression.association.Parenthese;
import ast.expression.binary.And;
import ast.expression.binary.Divide;
import ast.expression.binary.Equal;
import ast.expression.binary.Greater;
import ast.expression.binary.GreaterEqual;
import ast.expression.binary.Less;
import ast.expression.binary.LessEqual;
import ast.expression.binary.Minus;
import ast.expression.binary.Multiply;
import ast.expression.binary.NotEqual;
import ast.expression.binary.Or;
import ast.expression.binary.Plus;
import ast.expression.literal.BoolLiteral;
import ast.expression.literal.Identifier;
import ast.expression.literal.IntLiteral;
import ast.expression.literal.StrLiteral;
import ast.expression.unary.Negative;
import ast.expression.unary.Not;
import ast.expression.unary.Positive;
import ast.value.Bool;
import ast.value.Int;
import ast.value.Str;
import ast.value.Value;

public class Evaluator implements Visitor<Value> {

	public Evaluator() {
		
	}

	@Override
	public Value visit(Not node) {
		return node.getExpression().accept(this).not();
	}
	
	@Override
	public Value visit(Positive node) {
		return node.getExpression().accept(this).positive();
	}
	
	@Override
	public Value visit(Negative node) {
		return node.getExpression().accept(this).negative();
	}
	
	@Override
	public Value visit(Plus node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.plus(right);
	}

	@Override
	public Value visit(Minus node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.minus(right);
	}
	
	@Override
	public Value visit(Multiply node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.multiply(right);
	}
	
	@Override
	public Value visit(Divide node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.divide(right);
	}

	@Override
	public Value visit(And node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.and(right);
	}

	@Override
	public Value visit(Or node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.or(right);
	}

	@Override
	public Value visit(Equal node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.equal(right);
	}

	@Override
	public Value visit(NotEqual node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.notEqual(right);
	}

	@Override
	public Value visit(Greater node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.greater(right);
	}

	@Override
	public Value visit(GreaterEqual node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.greaterEqual(right);
	}

	@Override
	public Value visit(Less node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.less(right);
	}

	@Override
	public Value visit(LessEqual node) {
		Value left = node.getLeftExpression().accept(this);
		Value right = node.getRightExpression().accept(this);
		return left.lessEqual(right);
	}

	@Override
	public Value visit(Identifier node) {
		return null;
	}

	@Override
	public Value visit(IntLiteral node) {
		return new Int(node.getValue());
	}
	
	@Override
	public Value visit(BoolLiteral node) {
		return new Bool(node.getValue());
	}

	@Override
	public Value visit(StrLiteral node) {
		return new Str(node.getValue());
	}

	@Override
	public Value visit(Parenthese node) {
		return node.getExpression().accept(this);
	}

}

File: FelipezConde/QuestionnaireLanguage/AST/Evaluation/Evaluator.cs
using AST.Nodes.Interfaces;
using AST.Visitors;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AST.Nodes.Expression.Binary;
using AST.Nodes.Expression.Unary;
using AST.Nodes.Expression;
using AST.Nodes.Values;
using AST.Storage;
namespace AST.Evaluation
{
    public class Evaluator : BaseVisitor<Value>
    {
        ISymbolTable symbolTable;

        public Evaluator()
        {
            this.symbolTable = new SymbolTable();
        }

        public Value Evaluate(IExpression expression)
        {
            return expression.Accept(this);
        }

        public void AddValue(Id key, Value value)
        {
            this.symbolTable.AddValue(key, value);
        }

        public void UpdateValue(Id key, Value value)
        {
                this.symbolTable.SetUpdateValue(key, value);
        }

        public Value GetValue(Id key)
        {
            Value result = null;
            result = this.symbolTable.GetValue(key);
            return result;
        }


        public override Value Visit(Container node)
        {
            return node.Value.Accept(this);
        }

        #region Comparison
        public override Value Visit(And node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.BoolAnd((dynamic)right);
        }
        public override Value Visit(Or node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.BoolOr((dynamic)right);

        }
        public override Value Visit(Equal node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.BoolEqual((dynamic)right);
        }
        public override Value Visit(NotEqual node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.BoolNotEqual((dynamic)right);
        }
        public override Value Visit(GreaterThan node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.Greater((dynamic)right);
        }
        public override Value Visit(GreaterThanOrEqual node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.GreaterEqual((dynamic)right);
        }
        public override Value Visit(LessThan node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.Less((dynamic)right);
        }
        public override Value Visit(LessThanOrEqual node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.LessEqual((dynamic)right);
        }
        #endregion

        #region Unary Expressions
        public override Value Visit(Negate node)
        {

            Value value = node.Accept(this);
            return value.Negate();
        }
        public override Value Visit(Priority node)
        {
            return node.GetChildExpression().Accept(this);
        }
        #endregion

        #region Arithmetic

        public override Value Visit(Add node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.Add((dynamic)right);
        }

        public override Value Visit(Subtract node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.Substract((dynamic)right);
        }

        public override Value Visit(Multiply node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.Multiply((dynamic)right);
        }

        public override Value Visit(Divide node)
        {
            Value left = node.Left().Accept(this);
            Value right = node.Right().Accept(this);

            return left.Divide((dynamic)right);
        }


        #endregion

        #region Values
        public override Value Visit(Bool node)
        {
            return new Bool(node.GetValue());
        }

        public override Value Visit(Int node)
        {
            return new Int(node.GetValue());
        }

        public override Value Visit(Nodes.Values.String node)
        {
            return new Nodes.Values.String(node.GetValue());
        }
        #endregion

        public override Value Visit(Id node)
        {
            return this.GetValue(node);
        }
        
        public override Value Visit(Nodes.Labels.Label node)
        {
            return new Nodes.Values.String(node.Value);
        }

    }
}
